File: Round1/D/D.py
import numpy

M = 10 ** 9 + 7
cas = int(input())
for casi in range(cas):
    n, m = map(int, input().split())
    s = input()

    tn = n // 3
    if tn < m:
        print("0")
        continue

    f = numpy.ndarray([8, tn + 2], int)
    g = numpy.ndarray([8, tn + 2], int)
    f.fill(0)
    g.fill(0)
    f[0][0] = 1
    for si, c in enumerate(s):
        nf = si // 3 + 1 

        for i in range(1, 8):
            g[i][: nf] = 0
        for k in range(3):
            if c != chr(48 + k) and c != '?':
                continue
            for i in range(7):
                g[i | 1 << k][: nf] += f[i][: nf]
        for i in range(1, 8):
            g[i][: nf] %= M

        g[0][0] = 0
        g[0][1: nf + 1] = g[7][: nf]
        f, g = g, f
    ans = sum(sum(f[i][m: ]) for i in range(7)) % M
    print(ans)

File: Round1/D/D.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
)

func ReadLine(reader *bufio.Reader) string {
    line, _ := reader.ReadString('\n')
    return strings.TrimRight(line, "\n")
}

func ReadInt(reader *bufio.Reader) int {
    num, _ := strconv.Atoi(ReadLine(reader))
    return num
}

func ReadArray(reader *bufio.Reader) []int {
    line := ReadLine(reader);
    strs := strings.Split(line, " ")
    nums := make([]int, len(strs))
    for i, s := range strs {
        nums[i], _ = strconv.Atoi(s)
    }
    return nums
}

func main() {
    reader := bufio.NewReader(os.Stdin)
    cas := ReadInt(reader)

    const N int = 5000 + 10
    const M int = 1000000000 + 7
    for casi := int(0); casi < cas; casi++ {
        nm := ReadArray(reader)
        n, m := nm[0], nm[1]
        s := ReadLine(reader)

        tn := n / 3
        if tn < m {
            fmt.Println("0")
            continue
        }

        f := make([][]int, 8)
        g := make([][]int, 8)
        for i := 0; i < 8; i++ {
            f[i] = make([]int, tn + 2)
            g[i] = make([]int, tn + 2)
        }

        f[0][0] = 1
        for i := 1; i < 8; i++ {
            f[i][0] = 0
        }
        const ss string = "012"
        for si, c := range s {
            nf := si / 3 + 1

            for i := 1; i < 8; i++ {
                for j := 0; j < nf; j++ {
                    g[i][j] = 0
                }
            }
            var k uint
            for k = 0; k < 3; k++ {
                if c != rune(ss[k]) && c != '?' {
                    continue
                }
                for i := 0; i < 7; i++ {
                    ni := i | 1 << k
                    for j := 0; j < nf; j++ {
                        g[ni][j] = (g[ni][j] + f[i][j]) % M
                    }
                }
            }
            g[0][0] = 0
            for i := 0; i < nf; i++ {
                g[0][i + 1] = g[7][i]
            }
            f, g = g, f
        }
        var ans int = 0
        for i := 0; i < 7; i++ {
            for j := m; j <= tn; j++ {
                ans = (ans + f[i][j]) % M
            }
        }

        fmt.Println(ans)
    }
}
